File: SocialChallenge/Model/Persistent/ManagedObjects/JSCPost.m
//
//  JSCPost.m
//  SocialChallenge
//
//  Created by Javier Laguna on 06/08/2015.
//
//

#import "JSCPost.h"

#import "CDSServiceManager.h"
#import "NSManagedObjectContext+CDSRetrieval.h"

@implementation JSCPost

#pragma mark - Post

+ (instancetype)fetchPostWithId:(NSString *)postId
           managedObjectContext:(NSManagedObjectContext *)managedObjectContext
{
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"postId MATCHES %@", postId];
    
    return (JSCPost *)[managedObjectContext cds_retrieveFirstEntryForEntityClass:[JSCPost class]
                                                                       predicate:predicate];
}

+ (instancetype)fetchPostWithId:(NSString *)postId
{
    return [JSCPost fetchPostWithId:postId
               managedObjectContext:[CDSServiceManager sharedInstance].mainManagedObjectContext];
}

#pragma mark - UserName

- (NSString *)userName
{
    return [NSString stringWithFormat:@"%@ %@.", self.userFirstName, [self.userLastName substringWithRange:NSMakeRange(0,
                                                                                                               1)]];
}

@end

File: SocialChallenge/Model/Persistent/ManagedObjects/JSCPost.swift
//
//  JSCPost.swift
//  SocialChallenge
//
//  Created by Javier Laguna on 04/03/2016.
//
//

import Foundation
import CoreData
import CoreDataServices

class JSCPost: NSManagedObject {
    
    // MARK: Post
    
    /**
    Retrieves a JSCPOST from DB based on ID provided.
    
    - Parameter postId: ID of the post to be retrieved.
    - Parameter managedObjectContext: context that should be used to access persistent store.
    
    - Returns: JSCPOST instance or nil if POST can't be found.
    */
    class func fetchPostWithId(postId: String, managedObjectContext:NSManagedObjectContext) -> JSCPost? {
        
        let predicate = NSPredicate(format:"postId MATCHES %@", postId)
        
        var post: JSCPost?
        
        post = managedObjectContext.retrieveFirstEntry(JSCPost.self,
                                                       predicate: predicate) as? JSCPost
        
        return post
    }
    
    /**
     Retrieves a JSCPOST from DB based on ID provided, looking in the mainContext.
     
     - Parameter postId: ID of the post to be retrieved.
     
     - Returns: JSCPOST instance or nil if POST can't be found.
     */
    class func fetchPostWithId(postId: String) -> JSCPost? {
        
        var post: JSCPost?
        
        post = JSCPost.fetchPostWithId(postId, managedObjectContext: ServiceManager.sharedInstance.mainManagedObjectContext)
        
        return post
    }
    
    // MARK: UserName
    
    /**
    Convenient method to shape the user's name into the desired format.
    
    - Returns: user's Name.
    */
    func userName() -> String {
        
        let lastNameFirstLetter = self.userLastName![self.userLastName!.startIndex]
        
        return "\(self.userFirstName!) \(lastNameFirstLetter).";
    }
}
